core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -268,9 +270,9 @@ private boolean isEnoughVersionBasedAllocationsFound(IndexMetaData indexMetaData
      * Split the list of node shard states into groups yes/no/throttle based on allocation deciders
      */
     private NodesToAllocate buildNodesToAllocate(ShardRouting shard, RoutingAllocation allocation, List<NodeGatewayStartedShards> nodeShardStates) {
-        List<NodeGatewayStartedShards> yesNodeShards = new ArrayList<>();
-        List<NodeGatewayStartedShards> throttledNodeShards = new ArrayList<>();
-        List<NodeGatewayStartedShards> noNodeShards = new ArrayList<>();
+        List<Tuple<NodeGatewayStartedShards, Decision>> yesNodeShards = new ArrayList<>();
```

Comment:
If we only use Decision.Type, there is no need to for having all these tuples. yesNodeShards will only have decisions of type YES, etc.